src/Stack/Upgrade.hs

```diff
@@ -38,6 +41,9 @@ upgrade gitRepo mresolver = withSystemTempDirectory "stack-upgrade" $ \tmp' -> d
     tmp <- parseAbsDir tmp'
     mdir <- case gitRepo of
       Just repo -> do
+        remote <- liftIO $ readProcess "git" ["ls-remote", repo, "master"] []
+        let latestCommit = head . words $ remote
+        when (latestCommit == $(gitHash)) $ error "Already up-to-date, no upgrade required"
```

Comment:
Also: I don't think we should consider this an error condition, as that could break automated tooling that would simply want to know that it's using the newest stack. Can you instead just print a message and return `Nothing` like the non-Git case does?